library/core/src/panicking.rs

```diff
@@ -150,6 +150,20 @@ fn panic_bounds_check(index: usize, len: usize) -> ! {
     panic!("index out of bounds: the len is {len} but the index is {index}")
 }
 
+#[cold]
+#[cfg_attr(not(feature = "panic_immediate_abort"), inline(never))]
+#[track_caller]
+#[cfg_attr(not(bootstrap), lang = "panic_misaligned_pointer_dereference")] // needed by codegen for panic on misaligned pointer deref
+fn panic_misaligned_pointer_dereference(required: usize, found: usize) -> ! {
```

Comment:
Hmm... I think this should probably be a new variant of https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.AssertKind.html and use the https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.TerminatorKind.html#variant.Assert terminator

Reply:
I think it is. I think the problem is that machinery isn't capable of printing runtime values and needs to be manually wired up like this.

Reply:
Huh. I thought it printed slice indices and lengths already?

Reply:
Yeah, by going through its own helper in `library/core/src/panicking.rs` which works exactly like this.

Reply:
Oh 😄 carry on

Reply:
(I was secretly hoping you were going to tell me this whole mechanism can be obviated somehow because I personally find it quite convoluted)

Reply:
it is convoluted, but considering that we want this message to be #[cold] and never inlined, having explicit functions does seem reasonable, even if it's very annoying that we have to use a lang item per message.

Reply:
If we could mass an `enum` to the function we could have one lang item for all functions.

But keeping that in sync between libcore and the compiler is probably at least as annoying as having one function per message...